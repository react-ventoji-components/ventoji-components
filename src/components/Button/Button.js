import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

function calculateSize(size) {
  let fontSize;

  if (size === 'small') {
    fontSize = '12px';
  }

  if (size === 'medium') {
    fontSize = '14px';
  }

  if (size === 'large') {
    fontSize = '16px';
  }

  return fontSize;
}

const ButtonStyled = styled.button`
  color: ${(props) => props.color};
  font-size: ${(props) => props.fontSize};
  background-color: ${(props) => props.backgroundColor};
  padding: 10px 16px;
  cursor: pointer;
  font-weight: 700;
`;

/** Defautl button  */
function Button({ label, onClick, size, backgroundColor, color, type }) {
  return (
    <ButtonStyled
      type={type}
      onClick={onClick}
      fontSize={calculateSize(size)}
      backgroundColor={backgroundColor}
      color={color}
    >
      {label}
    </ButtonStyled>
  );
}

Button.propTypes = {
  /** Button Contents */
  label: PropTypes.string.isRequired,

  /** Function to call onClick */
  onClick: PropTypes.func.isRequired,

  /** Button type */
  type: PropTypes.oneOf(['button', 'submit']),

  /** Button size */
  size: PropTypes.oneOf(['small', 'medium', 'large']),

  /**  What background color to use */
  backgroundColor: PropTypes.string,

  /**  Font color */
  color: PropTypes.string,
};

Button.defaultProps = {
  color: 'black',
  size: 'small',
  label: 'click',
  type: 'button',
  onClick: () => {},
  backgroundColor: 'DFDFDF',
};

export default Button;
