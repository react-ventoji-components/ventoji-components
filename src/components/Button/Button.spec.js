import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Button from './Button';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

describe('Button component', () => {
  let button; // use shallow
  let tree; // use renderer
  let mockContext = 'small';

  beforeEach(() => {
    button = shallow(
      <Button label="button test" onClick={() => {}} size={mockContext} />,
    );

    tree = renderer.create(<Button />).toJSON();
  });

  it('Should be defined', () => {
    expect(Button).toBeDefined();
  });

  it('Should be button type', () => {
    expect(toJson(button)).toMatchSnapshot();

    expect(button.find({ type: 'button' })).toBeTruthy();
  });

  it('Should be submit type', () => {
    button = shallow(
      <Button
        label="button test"
        onClick={() => {}}
        size={mockContext}
        type="submit"
      />,
    );
    expect(toJson(button)).toMatchSnapshot();

    expect(button.find({ type: 'submit' })).toBeTruthy();
  });

  it('Should render correctly small size button', () => {
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('font-size', '12px');
    expect(tree).toHaveStyleRule('background-color', 'DFDFDF');
  });

  it('Should render correctly medium size button', () => {
    mockContext = 'medium';
    tree = renderer.create(<Button size={mockContext} />).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('font-size', '14px');
    expect(tree).toHaveStyleRule('background-color', 'DFDFDF');
  });

  it('Should render correctly large size button', () => {
    mockContext = 'large';
    tree = renderer.create(<Button size={mockContext} />).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('font-size', '16px');
    expect(tree).toHaveStyleRule('background-color', 'DFDFDF');
  });

  it('should call mock function when button is clicked', () => {
    const mockCallBack = jest.fn();

    const button = shallow(
      <Button onClick={mockCallBack} label="Ok" size={mockContext}></Button>,
    );

    //button.find('button').simulate('click');
    button.simulate('click');

    expect(mockCallBack.mock.calls.length).toEqual(1);
    expect(mockCallBack).toHaveBeenCalled();

    // console.log(button.find('button').props());
    // comment code is for a button (not styled-button)
  });
});
