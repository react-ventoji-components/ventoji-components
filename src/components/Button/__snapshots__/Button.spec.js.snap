// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Should be button type 1`] = `
<styled.button
  backgroundColor="DFDFDF"
  color="black"
  fontSize="12px"
  onClick={[Function]}
  type="button"
>
  button test
</styled.button>
`;

exports[`Button component Should be submit type 1`] = `
<styled.button
  backgroundColor="DFDFDF"
  color="black"
  fontSize="12px"
  onClick={[Function]}
  type="submit"
>
  button test
</styled.button>
`;

exports[`Button component Should render correctly large size button 1`] = `
.c0 {
  color: black;
  font-size: 16px;
  background-color: DFDFDF;
  padding: 10px 16px;
  cursor: pointer;
  font-weight: 700;
}

<button
  className="c0"
  color="black"
  fontSize="16px"
  onClick={[Function]}
  type="button"
>
  click
</button>
`;

exports[`Button component Should render correctly medium size button 1`] = `
.c0 {
  color: black;
  font-size: 14px;
  background-color: DFDFDF;
  padding: 10px 16px;
  cursor: pointer;
  font-weight: 700;
}

<button
  className="c0"
  color="black"
  fontSize="14px"
  onClick={[Function]}
  type="button"
>
  click
</button>
`;

exports[`Button component Should render correctly small size button 1`] = `
.c0 {
  color: black;
  font-size: 12px;
  background-color: DFDFDF;
  padding: 10px 16px;
  cursor: pointer;
  font-weight: 700;
}

<button
  className="c0"
  color="black"
  fontSize="12px"
  onClick={[Function]}
  type="button"
>
  click
</button>
`;
